### PR TITLE
Genquery cjsmeele master

### PR DESCRIPTION
**Pull request by d-w-moore**
Tests for [irods/irods_rule_engine_plugin_python#38]

### CODE REVIEW

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 1]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+                )
+                d0 = [ i for i in q ]
+                d1 = [ i for i in q.copy(options=Option.NO_DISTINCT) ]
+                callback.msiModAVUMetadata("-C",TestCollection,"rmw", "aa", "bb", "%")
+                b = repr([len(d0),len(d1)])
+                callback.writeLine('stdout',b)
+                callback.msiDeleteUnusedAVUs()
+                '''),
+            "no_distinct_",
+            lambda : eval(output) == [1,2]
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Use of possibly insecure function - consider using safer ast.literal_eval.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 2]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+        self.assertEqual(output.strip(), "[1,3]")
+
+    @unittest.skipIf(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'only applicable for python REP')
+    def test_query_objects(self):
+        if not self.full_test : return
+        os.mkdir(self.dir_for_coll)
+        interesting = [0,1,2,100,254,255,256,257,258,510,511,512,513,514]
+        octals = ['%04o'%x for x in interesting]
+        max_count = sorted(interesting)[-1]
+        max_octal = sorted(octals)[-1]
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Unused variable 'max_octal'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 3]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+                for j in {octals}:
+                    cond_string= "COLL_NAME = '{{c}}' and DATA_NAME < '{{ltparam}}'".format(c=TestCollection,ltparam=j)
+                    q = Query(callback,
+                              columns=["DATA_NAME"],
+                              conditions=cond_string)
+                    L.append(q.total_rows())
+                callback.writeLine("stdout", repr(L))
+                '''),
+            "total_rows_",
+            lambda : eval(output) == interesting
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Use of possibly insecure function - consider using safer ast.literal_eval.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 4]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+        max_octal = sorted(octals)[-1]
+        for x in range(max_count+1):
+             fname = os.path.join(self.dir_for_coll,'%04o'%x)
+             with open(fname,'wb') as fb:
+                 fb.write(os.urandom(x))
+        lib.execute_command ('tar -cf {} {}'.format(self.bundled_coll, self.dir_for_coll))
+        self.admin.assert_icommand('icd')
+        self.admin.assert_icommand('iput -f {}'.format(self.bundled_coll))
+        self.admin.assert_icommand('ibun -x {} .'.format(self.bundled_coll))
+        test_coll = self.test_admin_coll_path
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Unused variable 'test_coll'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 5]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+                        L.append(chunk)
+                    chunking_results[chunk_size] = L
+                check_Lresult = lambda chunkSize,L:\
+                  [int(x[0],8) for x in L] == list(range(0,n_results,chunkSize)) and \
+                  (0==len(L) or 1+((n_results-1)%chunkSize) == len(L[-1]))
+                #callback.writeLine('serverLog',pprint.pformat(chunking_results)) # - for DEBUG only
+                callback.writeLine('stdout',repr(list(check_Lresult(k,chunking_results[k]) for k in chunking_results)))
+                '''),
+            "chunked_queries_via_islice_",
+            lambda : list(filter(lambda result:result is False, eval(output))) == []
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Use of possibly insecure function - consider using safer ast.literal_eval.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 6]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+                          columns=["order_asc(DATA_NAME)"],
+                          conditions="DATA_NAME like '03%' and COLL_NAME = '{{c}}'".format(c=TestCollection)
+                )
+                q2 = q.copy(conditions="DATA_NAME like '05%' and COLL_NAME = '{{c}}'".format(c=TestCollection))
+                chained_queries = itertools.chain(q, q2)
+                compare_this = [x for x in ('%04o'%y for y in range({max_count}+1)) if x[:2] in ('03','05')]
+                callback.writeLine('stdout',repr([x for x in chained_queries] == compare_this and len(compare_this)==128))
+                '''),
+            "iter_chained_queries_",
+            lambda : eval(output) is True
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Use of possibly insecure function - consider using safer ast.literal_eval.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 7]
```diff
@@ -127,7 +133,195 @@ def setUp(self):
...
+                q = Query(callback,
+                          columns=["DATA_ID"],
+                          conditions="COLL_NAME = '{{c}}'".format(c=TestCollection))
+                q_iter = iter( q )
+                first_row = next(q_iter)
+                callback.writeLine('stdout',repr([1+len([row for row in q_iter]),
+                                                  q.copy().total_rows()]))
+                '''),
+            "copy_query_",
+            lambda : eval(output) == [ max_count+1 ]*2
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Use of possibly insecure function - consider using safer ast.literal_eval.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 8]
```diff
@@ -127,7 +134,194 @@ def setUp(self):
...
+        output, err, rc = self.admin.run_icommand("irule -F " + rule_file)
+        self.assertTrue(rc == 0, "icommand status ret = {r} output = '{o}' err='{e}'".format(r=rc,o=output,e=err))
+        self.assertEqual(output.strip(), "[1,3]")
+
+    @unittest.skipIf(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'only applicable for python REP')
+    def test_query_objects(self):
+        if not self.full_test : return
+        os.mkdir(self.dir_for_coll)
+        interesting = [0,1,2,100,254,255,256,257,258,510,511,512,513,514]
+        octals = ['%04o'%x for x in interesting]
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Unused variable 'octals'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py [hunk 9]
```diff
@@ -127,7 +134,194 @@ def setUp(self):
...
+        max_count = sorted(interesting)[-1]
+        for x in range(max_count+1):
+             fname = os.path.join(self.dir_for_coll,'%04o'%x)
+             with open(fname,'wb') as fb:
+                 fb.write(os.urandom(x))
+        lib.execute_command ('tar -cf {} {}'.format(self.bundled_coll, self.dir_for_coll))
+        self.admin.assert_icommand('icd')
+        self.admin.assert_icommand('iput -f {}'.format(self.bundled_coll))
+        self.admin.assert_icommand('ibun -x {} .'.format(self.bundled_coll))
+        test_coll = self.test_admin_coll_path
```
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Unused variable 'test_coll'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py (hunk 8 above)
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Unused variable 'octals'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py (hunk 9 above)
Comment (trel) [inline scripts/irods/test/test_prep_genquery_iterator.py]: Codacy Issue found: [Unused variable 'test_coll'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5016258)

File: scripts/irods/test/test_prep_genquery_iterator.py (hunk 8 above)
Comment (d-w-moore) [reply scripts/irods/test/test_prep_genquery_iterator.py]: the variable is used in Python formatting string later later in the file (line 210) as "{octals}"

File: scripts/irods/test/test_prep_genquery_iterator.py (hunk 9 above)
Comment (d-w-moore) [reply scripts/irods/test/test_prep_genquery_iterator.py]: the variable is used in Python formatting string later later in the file (line 196) as "{test_coll}"

**Pull request merged**